pennylane/optimize/spsa.py

```diff
@@ -0,0 +1,282 @@
+    ...    qml.CNOT(wires=[3, 1])
+    >>> def exp_val_circuit(params):
+    ...    circuit(params, range(dev.num_wires))
+    ...    return qml.expval(h2_ham)
+    >>> params = np.random.normal(0, np.pi, (num_qubits, 3), requires_grad=True)
+    >>> cost = qml.QNode(exp_val_circuit, dev)
+
+    Once constructed, the cost function can be passed directly to the
+    optimizer's ``step`` or ``step_and_cost`` function:
+
+    >>> max_iterations = 100
+    >>> opt = qml.SPSAOptimizer(maxiter=max_iterations)
+    >>> for n in range(max_iterations):
+    ...     params, energy = opt.step_and_cost(cost, params)
+
+    Example of hybrid classical-quantum workflow:
+
+    >>> dev = qml.device("default.qubit", wires=n_qubits)
+    >>> @qml.qnode(dev, interface=None)
+    >>> def layer_fn_spsa(inputs, weights):
+    ...     qml.AngleEmbedding(inputs, wires=range(n_qubits))
+    ...     qml.BasicEntanglerLayers(weights, wires=range(n_qubits))
+    ...     return [qml.expval(qml.PauliZ(wires=i)) for i in range(n_qubits)]
+
+    >>> opt = qml.SPSAOptimizer(maxiter=max_iterations)
+    >>> tensor_in = tf.Variable([0.27507603, 0.3453423])
+    >>> params = tf.Variable([[3.97507603, 2.00854038],
+    ...                       [3.12950603, 3.00854038],
+    ...                       [1.17907603, 1.10854038],
+    ...                       [0.97507603, 1.00854038],
+    ...                       [1.25907603, 0.40854088]])
+    >>> tensor_out = tf.Variable([0,0])
+
+    >>> init = tf.compat.v1.global_variables_initializer()
+    >>> with tf.compat.v1.Session() as sess:
+    ...     sess.run(init)
+    ...     for _ in range(max_iterations):
+    ...         # Take step
+    ...         params_a, layer_res = opt.step_and_cost(layer_fn_spsa,
+    ...                                np.tensor(tensor_in.eval(sess), requires_grad=False),
+    ...                                np.tensor(params.eval(sess)))
+    ...         params.assign(params_a[1], sess)
+    ...         tensor_out.assign(layer_res, sess)
+
+
+
+
+    Keyword Args:
+        maxiter (int): the maximum number of iterations expected to be performed.
+            Used to determine A, if A is not supplied, otherwise ignored.
+        alpha (float): A hyperparameter to calculate :math:`a_k=\frac{a}{(A+k+1)^\alpha}`
+            for each iteration. Its asymptotically optimal value is 1.0.
+        gamma (float): An hyperparameter to calculate :math:`c_k=\frac{c}{(k+1)^\gamma}`
+            for each iteration. Its asymptotically optimal value is 1/6.
+        c (float): A hyperparameter related to the expected noise. It should be
+            approximately the standard deviation of the expected noise on the cost function
+        A (float): The stability constant expected to be 10% of the maximum number
+            of expected iterations.
+        a (float): A hyperparameter expected to be small in noisy situations,
+            whose value could be :math:`\frac{mag(\Delta\theta)}{mag(g(\theta))}(A+1)^\alpha`
```

Comment:
Where does `\frac{mag(\Delta\theta)}{mag(g(\theta))}(A+1)^\alpha` come from? If it's not mentioned in the Overview in the docstring, then it's worth citing another source.

Reply:
I've added the reference in which Spall makes some recommendations about constants and gains